modnet/models.py

```diff
@@ -554,79 +566,904 @@ class OR only return the most probable class.
+
+        Parameters:
+            test_data: A featurized and feature-selected `MODData`
+                object containing the descriptors used in training.
+            return_prob: For a classification tasks only: whether to return the probability of each
+                class OR only return the most probable class.
+            return_unc: whether to return the standard deviation as a second dataframe
+
+        Returns:
+            A `pandas.DataFrame` containing the predicted values of the targets.
+            If return_unc=True, two `pandas.DataFrame` : (predictions,std) containing the predicted values of the targets and
+             the standard deviations of the epistemic uncertainty.
+
+
+        """
+        # prevents Nan predictions if some features are inf
+        x = test_data.get_featurized_df().replace([np.inf, -np.inf, np.nan], 0)[
+            self.optimal_descriptors[:self.n_feat]
+        ].values
+
+        # Scale the input features:
+        x = np.nan_to_num(x)
+        if self._scaler is not None:
+            x = self._scaler.transform(x)
+            x = np.nan_to_num(x)
+
+        all_predictions = []
+
+        for i in range(1000):
+            p = self.model.predict(x)
+            if len(self.targets_flatten) == 1:
+                p = np.array([p])
+            all_predictions.append(p)
+
+        p_dic = {}
+        unc_dic = {}
+        for i, name in enumerate(self.targets_flatten):
+            if self.num_classes[name] >= 2:
+                if return_prob:
+                    preds = np.array([pred[i] for pred in all_predictions])
+                    probs = preds/(preds.sum(axis=-1)).reshape((-1, 1))
+                    mean_prob = probs.mean()
+                    std_prob = probs.std()
+                    for j in range(mean_prob.shape[-1]):
+                        p_dic['{}_prob_{}'.format(name, j)] = mean_prob[:,j]
+                        unc_dic['{}_prob_{}'.format(name, j)] = std_prob[:,j]
+                else:
+                    p_dic[name] = np.argmax(np.array([pred[i] for pred in all_predictions]).mean(axis=0), axis=1)
+                    unc_dic[name] = np.max(np.array([pred[i] for pred in all_predictions]).mean(axis=0), axis=1)
+            else:
+                mean_p = np.array([pred[i] for pred in all_predictions]).mean(axis=0)
+                std_p = np.array([pred[i] for pred in all_predictions]).std(axis=0)
+                p_dic[name] = mean_p[:,0]
+                unc_dic[name] = std_p[:,0]
+
+        predictions = pd.DataFrame(p_dic)
+        unc = pd.DataFrame(unc_dic)
+
+        predictions.index = test_data.structure_ids
+        unc.index = test_data.structure_ids
+
+        if return_unc:
+            return predictions, unc
+        else:
+            return predictions
+
+    def fit_preset(*args,**kwargs):
+        '''Not implemented'''
+
+        raise RuntimeError(
+            "Not implemented."
+        )
```

Comment:
We should make an abstract `BaseMODNetModel` that highlights what functions are expected, but until then:

```suggestion
    def fit_preset(*args,**kwargs):
        raise NotImplementedError
```

Reply:
Saved for next PR